csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -136,14 +136,17 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
             var inputNames = new string[inputs.Count];
             var inputTensors = new IntPtr[inputs.Count];
             var pinnedBufferHandles = new System.Buffers.MemoryHandle[inputs.Count];
+            var disposeOnnxValues = new bool[inputs.Count];
 
             int inputIndex = 0;
             foreach (var input in inputs)
             {
                 inputNames[inputIndex] = input.Name;
 
                 // create Tensor from the input if feasible, else throw notsupported exception for now
-                input.ToNativeOnnxValue(out inputTensors[inputIndex], out pinnedBufferHandles[inputIndex]);
+                input.ToNativeOnnxValue(out inputTensors[inputIndex], 
+                                        out pinnedBufferHandles[inputIndex], 
+                                        out disposeOnnxValues[inputIndex]);
```

Comment:
Should remove array disposeOnnxValues if possible.

For example, In the 'finally' block below, if input type is NamedOnnxValue  then dispose pinnedBufferHandles, otherwise don't dispose. This bypasses need for disposeOnnxValues array.

Reply:
Is there a specific reason having disposeOnnxValues array might be an issue ? Let's say there is a new derived class in the future, doesn't it make things easier if we keep a boolean flag indicating if we need to dispose the underlying onnxValue rather than have the finally block handle the logic based on the type(s). It feels like the current approach is more easily extensible.

Reply:
It simplifies the signature of ToNativeOnnxValue(). There's one less parameter to pass and less variables/operations to keep track of. Not sure if future extensibility is needed here -- additional parameters/variables can always be added in future if needed.

Reply:
Removed the extra bool out parameter. In InferenceSession.cs, conditionally dispose off the native OrtValue if the instance is strictly of type NamedOnnxValue only.